### PR TITLE
Port grouping redo/v17

### PR DESCRIPTION
**Pull request by inashivb**
Previous PR: https://github.com/OISF/suricata/pull/10482

Changes since v16:
- some semantic fixes
- optimization of whitelisting (ref: https://github.com/OISF/suricata/pull/10482#issuecomment-1956900874)
- 
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) tickets:
- https://redmine.openinfosecfoundation.org/issues/6792
- https://redmine.openinfosecfoundation.org/issues/6795

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1662

### CODE REVIEW

**Review by inashivb — COMMENTED** [review]
This offers a performance upgrade of about a minute on my laptop compared to previous PR.
This PR: `2m44.632s` vs
v16: `3m42.618s` vs
master: `13m5s` 

There is still a bit more scope of optimization as commented inline. Lmk what do you think @victorjulien

File: src/detect-engine-build.c [hunk 1]
```diff
@@ -1125,9 +1126,376 @@ static int RuleSetWhitelist(Signature *s)
...
+                insert = insert->next;
+            }
+            tmp->next = insert->next;
+            insert->next = tmp;
+
+            if (insert == cur) {
+                cur = tmp;
+            }
+        }
+    }
```
Comment (inashivb) [inline src/detect-engine-build.c:1176]: Could use a better sorting algorithm than insertion sort. This is quadratic complexity.

File: src/detect-engine-build.c [hunk 2]
```diff
@@ -1125,9 +1126,376 @@ static int RuleSetWhitelist(Signature *s)
...
+{
+    for (uint8_t i = 0; i < RULE_GROUP_MAX; i++) {
+        if (list_group[i] != NULL) {
+            if (*tmplist == NULL) {
+                *tmplist = list_group[i];
+            } else {
+                (*tmplist)->last->next = list_group[i];
+            }
+            DetectPort *tmp = NULL;
+            for (tmp = list_group[i]; tmp->next != NULL; tmp = tmp->next)
```
Comment (inashivb) [inline src/detect-engine-build.c:1206]: Could store the tail of the list and avoid looping all through it.

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 101 | 111 | 109.9% | 

Pipeline 18705

**Comment by victorjulien** [comment]
> This offers a performance upgrade of about a minute on my laptop compared to previous PR. This PR: `2m44.632s` vs v16: `3m42.618s` vs master: `13m5s`
> 
> There is still a bit more scope of optimization as commented inline. Lmk what do you think @victorjulien

I don't really see an improvement here. Maybe 10s or so.

**Comment by victorjulien** [comment]
> > This offers a performance upgrade of about a minute on my laptop compared to previous PR. This PR: `2m44.632s` vs v16: `3m42.618s` vs master: `13m5s`
> > There is still a bit more scope of optimization as commented inline. Lmk what do you think @victorjulien
> 
> I don't really see an improvement here. Maybe 10s or so.

Perf looks almost the same
image

File: src/detect-engine-build.c (hunk 1 above)
Comment (inashivb) [reply src/detect-engine-build.c:1176]: Done in https://github.com/OISF/suricata/pull/10497 by Victor

File: src/detect-engine-build.c (hunk 2 above)
Comment (inashivb) [reply src/detect-engine-build.c:1206]: No longer needed.

**Comment by inashivb** [comment]
Replaced w https://github.com/OISF/suricata/pull/10505

**Pull request closed without merging**